src/Symfony/Component/String/AbstractUnicodeString.php

```diff
@@ -0,0 +1,472 @@
+    }
+
+    public function padStart(int $length, string $padStr = ' '): parent
+    {
+        if ('' === $padStr || !preg_match('//u', $padStr)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $pad = clone $this;
+        $pad->string = $padStr;
+
+        return $this->pad($length, $pad, STR_PAD_LEFT);
+    }
+
+    public function replaceMatches(string $fromPattern, $to): parent
+    {
+        if ($this->ignoreCase) {
+            $fromPattern .= 'i';
+        }
+
+        if (\is_array($to) || $to instanceof \Closure) {
+            if (!\is_callable($to)) {
+                throw new \TypeError(sprintf('Argument 2 passed to %s::replaceMatches() must be callable, array given.', \get_class($this)));
+            }
+
+            $replace = 'preg_replace_callback';
+            $to = static function (array $m) use ($to): string {
+                $to = $to($m);
+
+                if ('' !== $to && (!\is_string($to) || !preg_match('//u', $to))) {
+                    throw new InvalidArgumentException('Replace callback must return a valid UTF-8 string.');
+                }
+
+                return $to;
+            };
+        } elseif ('' !== $to && !preg_match('//u', $to)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        } else {
+            $replace = 'preg_replace';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (null === $string = $replace($fromPattern.'u', $to, $this->string)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        $str = clone $this;
+        $str->string = $string;
+
+        return $str;
+    }
+
+    /**
+     * @return static
+     */
+    public function slug(string $separator = '-'): self
+    {
+        return $this
+            ->ascii()
+            ->replace('@', $separator.'at'.$separator)
+            ->replaceMatches('/[^A-Za-z0-9]++/', $separator)
+            ->trim($separator);
+    }
+
+    public function snake(): parent
+    {
+        $str = $this->camel()->title();
+        $str->string = mb_strtolower(preg_replace(['/(\p{Lu}+)(\p{Lu}\p{Ll})/u', '/([\p{Ll}0-9])(\p{Lu})/u'], '\1_\2', $str->string), 'UTF-8');
+
+        return $str;
+    }
+
+    public function title(bool $allWords = false): parent
+    {
+        $str = clone $this;
+
+        if ($allWords) {
+            $str->string = preg_replace_callback('/\b./u', static function ($m) {
```

Comment:
```suggestion
            $str->string = preg_replace_callback('/\b./u', static function ($m): string {
```
Add another type hint safety guard on anonymous function